circuit_setup/mdl-tools/src/bin/prepare-prover-input.rs

```diff
@@ -0,0 +1,481 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// This program parses a CBOR-encoded mDL and tests extracting the required information
+// to generate the ZK circuit inputs.
+//
+// Usage:
+//    prepare-prover-input --config <config> --mdl <mdl> --prover_inputs <prover_inputs>
+// where
+//    <config> is a JSON file containing the configuration
+//    <mdl> is a CBOR-encoded mDL
+//    <prover_inputs> is the output JSON file containing the prover inputs
+//
+// Notes:
+//    - The mDL file can be generated by the mdl-gen program
+//    - To test: cargo run --bin mdl-test -- --config ../inputs/mdl1/config.json --mdl ../inputs/mdl1/mdl.cbor --prover_inputs ../generated_files/mdl1/prover_inputs.json
+
+use anyhow::Result;
+use clap::Parser;
+use coset::cbor::Value;
+use coset::Label;
+use isomdl::cbor;
+use isomdl::definitions::helpers::{ByteStr, NonEmptyVec};
+use isomdl::definitions::issuer_signed::IssuerSignedItemBytes;
+use isomdl::definitions::x509::x5chain::X5CHAIN_COSE_HEADER_LABEL;
+use isomdl::definitions::x509::X5Chain;
+use isomdl::definitions::{DigestAlgorithm, DigestId, Mso};
+use isomdl::issuance::mdoc::Mdoc;
+use p256::ecdsa::{Signature, VerifyingKey};
+use p256::pkcs8::EncodePublicKey;
+use p256::NistP256;
+use serde_json::Map;
+use sha2::{Digest, Sha256};
+use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
+use num_bigint::BigUint;
+use num_traits::{Num, Zero, One, ToPrimitive};
+use chrono::{DateTime, Datelike, FixedOffset, NaiveDate, NaiveDateTime, TimeZone, Utc};
+use hex;
+use std::str;
+use std::collections::BTreeMap;
+
+static MDL_DOCTYPE: &str = "org.iso.18013.5.1.mDL";
+static ISO_MDL_NAMESPACE: &str = "org.iso.18013.5.1";
+static AAMVA_MDL_NAMESPACE: &str = "org.iso.18013.5.1.aamva";
+static SUPPORTED_NAMESPACES: [&str; 2] = [ISO_MDL_NAMESPACE, AAMVA_MDL_NAMESPACE];
+const CIRCOM_ES256_LIMB_BITS: usize = 43;
+
+// Define the issuer's timezone as PST (UTC-8). Change this if needed (TODO: move to config)
+const ISSUER_TIMEZONE_OFFSET_HOURS: i32 = -8; // US West Coast / PST (UTC-8)
+
+#[derive(Parser, Debug)]
+#[command(author, version, about)]
+struct Args {
+    /// JSON file containing the config
+    #[arg(short = 'c', long = "config")]
+    config: String,
+
+    /// JSON file containing the mDL ISO claims
+    #[arg(short = 'm', long = "mdl")]
+    mdl: String,
+
+    /// output JSON file containing the prover inputs
+    #[arg(short = 'i', long = "prover_inputs")]
+    prover_inputs: String,
+}
+
+fn sha256_padding(prepad_m: &[u8]) -> Vec<u8> {
```

Comment:
Is some of this code duplicated from https://github.com/microsoft/crescent-credentials/blob/3397e0052e04fbc57ff4768f6ff0ede087433111/creds/src/prep_inputs.rs#L485

Reply:
Yeah, mostly copied. I can import it instead.

Reply:
There is a lot of overlap between the two files. We'll need to merge them down at some point.